private/pkg/manifest/storage.go

```diff
@@ -0,0 +1,158 @@
+func (o *manifestBucketObject) Close() error               { return o.file.Close() }
+
+// NewFromBucket creates a manifest from a storage bucket, with all its digests
+// in DigestTypeShake256.
+func NewFromBucket(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+) (*Manifest, error) {
+	m := New()
+	digester, err := NewDigester(DigestTypeShake256)
+	if err != nil {
+		return nil, err
+	}
+	if walkErr := bucket.Walk(ctx, "", func(info storage.ObjectInfo) (retErr error) {
+		path := info.Path()
+		obj, err := bucket.Get(ctx, path)
+		if err != nil {
+			return err
+		}
+		defer func() { retErr = multierr.Append(retErr, obj.Close()) }()
+		digest, err := digester.Digest(obj)
+		if err != nil {
+			return err
+		}
+		if err := m.AddEntry(path, *digest); err != nil {
+			return err
+		}
+		return nil
+	}); walkErr != nil {
+		return nil, walkErr
+	}
+	return m, nil
+}
+
+// NewBucket takes a manifest and an array of blobs and builds a readable storage bucket that
+// contains the files in the manifest.
+func NewBucket(m Manifest, blobs []Blob) (storage.ReadBucket, error) {
+	digestToBlobs := make(map[string]Blob, len(blobs))
+	for _, b := range blobs {
+		digest := b.Digest().String()
+		_, alreadyPresent := digestToBlobs[digest]
+		if alreadyPresent {
+			return nil, fmt.Errorf("blob with digest %q duplicated", digest)
+		}
+		_, presentInManifest := m.PathsFor(digest)
+		if !presentInManifest {
+			return nil, fmt.Errorf("blob with digest %q is not present in manifest", digest)
+		}
+		digestToBlobs[digest] = b
+	}
+	pathToBlob := make(map[string]Blob, len(blobs))
+	for _, path := range m.Paths() {
+		pathDigest, ok := m.DigestFor(path)
+		if !ok {
+			return nil, fmt.Errorf("path %q not present in manifest", path)
+		}
+		blob, ok := digestToBlobs[pathDigest.String()]
+		if !ok {
+			return nil, fmt.Errorf("manifest path %q with digest %q has no associated blob", path, pathDigest.String())
+		}
+		pathToBlob[path] = blob
+	}
+	return &manifestBucket{
+		manifest:   m,
+		pathToBlob: pathToBlob,
+	}, nil
+}
+
+func (m *manifestBucket) Get(ctx context.Context, path string) (storage.ReadObjectCloser, error) {
+	blob, ok := m.pathToBlob[path]
+	if !ok {
+		return nil, storage.NewErrNotExist(path)
+	}
+	file, err := blob.Open(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &manifestBucketObject{
+		path: path,
+		file: file,
+	}, nil
+}
+
+func (m *manifestBucket) Stat(ctx context.Context, path string) (storage.ObjectInfo, error) {
+	_, ok := m.pathToBlob[path]
+	if !ok {
+		return nil, storage.NewErrNotExist(path)
+	}
+	// storage.ObjectInfo only requires path
+	return &manifestBucketObject{path: path}, nil
+}
+func (m *manifestBucket) Walk(ctx context.Context, prefix string, f func(storage.ObjectInfo) error) error {
+	prefix, err := storageutil.ValidatePrefix(prefix)
+	if err != nil {
+		return err
+	}
+	paths := m.manifest.Paths()
+	sort.Slice(paths, func(i, j int) bool {
+		return paths[i] < paths[j]
+	})
```

Comment:
I don't think paths are walked in any sort order. At least it's not defined by ReadBucket.Walk and I can't imagine another implementation trying to force such (e.g. storageos).

Reply:
I found this behavior in `storagemem` [here](https://github.com/bufbuild/buf/blob/61c2f49aa335dce8bc2311c116b84b61999c4ee5/private/pkg/storage/storagemem/bucket.go#L63-L70).

Reply:
Removed